README.md
# Home Assistant Integration for Moat Temperature/Humidity BLE Sensors

A custom component for [Home Assistant](https://www.home-assistant.io) that listens for the BLE (Bluetooth Low Energy) advertising packets broadcast by Moat Bluetooth Thermometer/Hygrometers.

[![hacs_badge](https://img.shields.io/badge/HACS-Custom-orange.svg?style=plastic)](https://github.com/custom-components/hacs)
[![Language grade: Python](https://img.shields.io/lgtm/grade/python/g/SteveOnorato/moat_temp_hum_ble.svg?logo=lgtm&logoWidth=18&style=plastic)](https://lgtm.com/projects/g/SteveOnorato/moat_temp_hum_ble/context:python)
[![Total alerts](https://img.shields.io/lgtm/alerts/g/SteveOnorato/moat_temp_hum_ble.svg?logo=lgtm&logoWidth=18&style=plastic)](https://lgtm.com/projects/g/SteveOnorato/moat_temp_hum_ble/alerts/)
## Supported Devices
* [Moat S2](https://www.amazon.com/dp/B08DK739F5)
* Select Govee BLE Sensors


## Installation


**1. Install the custom component:**

- The easiest way is to install it with [HACS](https://hacs.xyz/).
  - First install [HACS](https://hacs.xyz/) if you don't have it yet.
  - Click on "HACS" in the Home Assistant UI sidebar to display the "Home Assistant Community Store" page.
  - Click "Integrations".
  - Click the vertical ellipsis in the top-right, choose "Custom repositories"
    - Choose "Integration" as the Category.
    - Paste in the custom repository URL: https://github.com/SteveOnorato/moat_temp_hum_ble
    - Click "ADD".
    - Click "X" to close the dialog.
  - Click "INSTALL" under the new "Moat BLE Temperature/Humidity Sensor" item.

- Alternatively, you can install it manually. Just copy & paste the content of the `moat_temp_hum_ble/custom_components` folder into your `config/custom_components` directory.
     As example, you will get the `sensor.py` file in the following path: `/config/custom_components/moat_temp_hum_ble/sensor.py`.

*NOTE:* the following instructions about setting device permissions are an edge case for a very specific set up.  (If you do not understand, do not worry about it).
- If running Home Assistant without root access, the [Bleson](https://github.com/TheCellule/python-bleson) Python library used for accessing Bluetooth requires the following permissions applied to the Python 3 binary. If using a virtual environment for HA, this binary will be in the virtual environment path.

     *NOTE*: Replace "path" with the path to the Python3 binary (example: /srv/homeassistant/bin/python3)
     ```
     sudo setcap cap_net_raw,cap_net_admin+eip $(eval readlink -f path)
     ```
**2. Restart Home Assistant:**

- Restart Home Assistant to load the integration.
  
  You can use the UI: Configuration / Server Controls / RESTART.
- Make sure you do this **before** modifying configuration.yaml, otherwise the configuration won't be recognized and you'll get an error when clicking RESTART.

**3. Add the platform to your configuration.yaml file (see [below](#configuration-variables))**

**4. Restart Home Assistant:**

- A second restart is required to load the configuration. Within a few minutes, the sensors should be added to your home-assistant automatically (wait at least period_secs).

**5. If the entities are still not displaying data, a restart of the host device may be required.**


### Configuration Variables

In **configuration.yaml**, specify the sensor platform `moat_temp_hum_ble` and a list of devices with unique MAC address.

#### Simple configuration example:
*NOTE*: device name is optional.  If not provided, devices will be labeled using the MAC address.
```
sensor:
  - platform: moat_temp_hum_ble
    moat_devices:
      - mac: "A4:C1:38:A1:A2:A3"
        name: Bedroom
      - mac: "A4:C1:38:B1:B2:B3"
      - mac: "A4:C1:38:C1:C2:C3"
        name: Kitchen
```

There are multiple ways to learn the MAC addresses for your Bluetooth devices.
##### Windows
* Use https://www.microsoft.com/en-us/p/bluetooth-le-explorer/9n0ztkf1qd98?activetab=pivot:overviewtab (thanks, @iamhueman)
##### macOS
* Use packetlogger.app from "Additional Tools for XCode" https://developer.apple.com/download/more/?=packetlogger (thanks, @tronicdude)
##### Home Assistant Operating System (formerly HassOS):
* Enable the SSH & Web Terminal Add-on under Supervisor -> Dashboard (see https://community.home-assistant.io/t/home-assistant-community-add-on-ssh-web-terminal/33820).
  * You likely need to disable 'Protection mode' for SSH & Web Terminal and restart the add-on (to enable docker commands)
* Then, use the Web Terminal to run the following:
```
docker exec -it $(docker ps -f name=homeassistant -q) bash
hciconfig hci0 down
hciconfig hci0 up
hcitool -i hci0 lescan | grep -i 'Govee\|GVH\|Moat'
```
* Leave this running for a bit and it will display the matching devices as it hears from them.
* Hit Ctrl+C when done.
* You might want to re-enable 'Protection mode' for SSH & Web Terminal at this point.

#### Additional platform configuration options
| Option | Type |Default Value | Description |  
| -- | -- | -- | -- |
| `report_fahrenheit` | Boolean | `False` | True for Fahrenheit, False for Celsius. |
| `rounding`| Boolean | `True` | Enable/disable rounding of the measurements reported to Home Assistant.  (Either way, we always use maximum precision while collecting the data points to be averaged each `period`.)  NOTE: By default, this will round as Celsius, but if 'report_fahrenheit' is True, this will round as Fahrenheit.  If the Home Assistant UI converts to the other unit (due to global preference), you may see more decimal places than expected. |
| `decimals` | positive integer | `2`| Number of decimal places to round to (only if `rounding` is enabled). |
| `period_secs` | positive integer | `60` | The amount of time, in seconds, during which the sensor readings are collected before reporting the average to Home Assistant. The devices broadcast roughly once per second, so this limits the amount of mostly duplicate data stored in Home Assistant's database. |
| `log_spikes` | Boolean | `True` | Puts information about each erroneous spike in the Home Assistant log. |
| `update_when_unavailable` | Boolean | `True` | Determines the behavior when no measurements are received during a reporting period. If True, sets the temperature/humidity/RSSI/battery states to "unavailable". If False, the temperature/humidity/RSSI/battery states will not be updated (this will cause Home Assistant to keep showing the previous value, even though it may be out-of-date). |
| `use_median` | Boolean  | `False` | Use median as sensor output instead of mean (helps with "spiky" sensors). Either way, both the median and the mean values are present in the device state attributes (which can be used to make Template Sensors). |
| `temp_range_min_celsius` | float | `-45.0` | Can set the lower bound of reasonable measurements, in Celsius (even if `report_fahrenheit` is enabled).  Temperature measurements lower than this will be discarded. |
| `temp_range_max_celsius` | float | `70.0` | Can set the upper bound of reasonable measurements, in Celsius (even if `report_fahrenheit` is enabled).  Temperature measurements higher than this will be discarded. |
| `temperature_entities` | Boolean  | `True` | Can disable this if you don't want the temperature entities for the sensor devices. |
| `humidity_entities` | Boolean  | `True` | Can disable this if you don't want the humidity entities for the sensor devices. |
| `battery_entities` | Boolean  | `False` | Can enable this if you want a separate entity to track the battery percentage for each sensor device. |
| `num_samples_entities` | Boolean  | `False` | Can enable this if you want a separate entity to track the number of samples received each period for each sensor device. |
| `rssi_entities` | Boolean  | `False` | Can enable this if you want a separate entity to track the RSSI for each sensor device. |
| `hci_device`| string | `hci0` | HCI device name used for scanning.  May need to be changed if you have multiple Bluetooth adapters connected. |
| `govee_devices` | list of objects | None | Same format as `moat_devices`, but supports (BLE mode only) for Govee sensors H5051, H5072, H5074, H5075, H5101, H5102, and H5177.  I use this because the "python-bleson" library used here only supports 1 scan at a time, so this integration can't successfully run at the same time as https://github.com/Home-Is-Where-You-Hang-Your-Hack/sensor.goveetemp_bt_hci |

#### Additional device configuration options
| Option | Type |Default Value | Description |  
| -- | -- | -- | -- |
| `calibrate_temp` | float | `0.0` | Add this amount to each temperature measurement for this device (in degrees Fahrenheit or Celsius, depending on the "report_fahrenheit" setting). |
| `calibrate_humidity` | float | `0.0` | Add this amount to each humidity measurement for this device (in %). |

#### Full configuration example:
```
sensor:
  - platform: moat_temp_hum_ble
    report_fahrenheit: True
    rounding: True
    decimals: 1
    period_secs: 60
    log_spikes: True
    update_when_unavailable: True
    use_median: False
    temp_range_min_celsius: -45.0
    temp_range_max_celsius: 70.0
    temperature_entities: True
    humidity_entities: True
    battery_entities: False
    num_samples_entities: True
    rssi_entities: False
    hci_device: hci0
    moat_devices:
      - mac: "DE:49:A1:A2:A3:A4"
        name: Deep Freezer
        calibrate_temp: -0.55
        calibrate_humidity: -3.0
      - mac: "DE:49:C1:C2:C3:C4"
        name: Kitchen
        calibrate_temp: 2.2
    govee_devices:
      - mac: "A4:C1:38:A1:A2:A3"
        name: Bedroom
        calibrate_temp: 0.4
```

### Debugging
To enable logging for this component, add the following to configuration.yaml:
```
logger:
  logs:
    custom_components.moat_temp_hum_ble: debug
```


## Credits
  This was forked from https://github.com/Home-Is-Where-You-Hang-Your-Hack/sensor.goveetemp_bt_hci, which itself was based on [custom-components/sensor.mitemp_bt](https://github.com/custom-components/sensor.mitemp_bt).
  So, a big thank you to [@Thrilleratplay](https://community.home-assistant.io/u/thrilleratplay), [@tsymbaliuk](https://community.home-assistant.io/u/tsymbaliuk), and [@Magalex](https://community.home-assistant.io/u/Magalex)!
  


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## 0.1.2
  - Reports detected model of each device as an entity attribute.
  - Better handling of devices that don't conform to the BLE specification.
## 0.1.1
  - Add fix for temperatures below 0 Celsius (only affects Govee H5072/5075/5101/5102).
## 0.1.0
  - **Initial Release**


custom_components/moat_temp_hum_ble/sensor.py
"""Integration with Home Assistant (data types, platform setup, and update loop).

Note that this file MUST be named sensor.py to match the top level name in
configuration.yaml.
"""
import logging
from dataclasses import dataclass
from datetime import timedelta
from typing import List, Optional, Dict, Any, Sequence, Union

import voluptuous as vol  # type: ignore

import homeassistant.helpers.config_validation as cv  # type: ignore
import homeassistant.util.dt as dt_util  # type: ignore
from bleson import get_provider, Adapter  # type: ignore
from bleson.core.hci.constants import EVT_LE_ADVERTISING_REPORT  # type: ignore
from bleson.core.hci.type_converters import hex_string  # type: ignore
from bleson.core.types import BDAddress  # type: ignore
from homeassistant.components import sensor
from homeassistant.components.sensor import PLATFORM_SCHEMA  # type: ignore
from homeassistant.const import (  # type: ignore
    DEVICE_CLASS_TEMPERATURE,
    DEVICE_CLASS_HUMIDITY,
    DEVICE_CLASS_BATTERY,
    DEVICE_CLASS_SIGNAL_STRENGTH,
    TEMP_CELSIUS,
    TEMP_FAHRENHEIT,
    ATTR_BATTERY_LEVEL,
)
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.entity import Entity  # type: ignore
from homeassistant.helpers.event import track_point_in_utc_time  # type: ignore
from .const import (
    CONF_DEVICE_MAC,
    CONF_DEVICE_NAME,
    CONF_DEVICE_CALIBRATE_TEMP,
    CONF_DEVICE_CALIBRATE_HUMIDITY,
    CONF_REPORT_FAHRENHEIT,
    DEFAULT_REPORT_FAHRENHEIT,
    CONF_ROUNDING,
    DEFAULT_ROUNDING,
    DEFAULT_DECIMALS,
    CONF_DECIMALS,
    CONF_PERIOD_SECS,
    DEFAULT_PERIOD,
    CONF_LOG_SPIKES,
    DEFAULT_LOG_SPIKES,
    CONF_UPDATE_WHEN_UNAVAILABLE,
    DEFAULT_UPDATE_WHEN_UNAVAILABLE,
    CONF_USE_MEDIAN,
    CONF_TEMP_RANGE_MIN_CELSIUS,
    CONF_TEMP_RANGE_MAX_CELSIUS,
    DEFAULT_TEMP_RANGE_MAX,
    CONF_TEMPERATURE_ENTITIES,
    DEFAULT_TEMPERATURE_ENTITIES,
    CONF_HUMIDITY_ENTITIES,
    CONF_BATTERY_ENTITIES,
    DEFAULT_BATTERY_ENTITIES,
    CONF_RSSI_ENTITIES,
    DEFAULT_RSSI_ENTITIES,
    CONF_NUM_SAMPLES_ENTITIES,
    DEFAULT_NUM_SAMPLES_ENTITIES,
    CONF_GOVEE_DEVICES,
    CONF_HCI_DEVICE,
    CONF_MOAT_DEVICES,
    DEFAULT_HCI_DEVICE,
    DEFAULT_TEMP_RANGE_MIN,
    DEFAULT_USE_MEDIAN,
    DeviceBrand,
    DEFAULT_DEVICE_CALIBRATE_TEMP,
    DEFAULT_DEVICE_CALIBRATE_HUMIDITY,
)
from .const import DEFAULT_HUMIDITY_ENTITIES
from .sensor_device import SensorDevice, CreateDeviceParams
from .temp_hum_advertisement_parser import parse_ble_advertisement

###############################################################################

_LOGGER = logging.getLogger(__name__)

DEVICES_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_DEVICE_MAC): cv.string,
        vol.Optional(CONF_DEVICE_NAME): cv.string,
        vol.Optional(CONF_DEVICE_CALIBRATE_TEMP): float,
        vol.Optional(CONF_DEVICE_CALIBRATE_HUMIDITY): float,
    }
)

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
    {
        vol.Optional(
            CONF_REPORT_FAHRENHEIT, default=DEFAULT_REPORT_FAHRENHEIT
        ): cv.boolean,
        vol.Optional(CONF_ROUNDING, default=DEFAULT_ROUNDING): cv.boolean,
        vol.Optional(CONF_DECIMALS, default=DEFAULT_DECIMALS): cv.positive_int,
        vol.Optional(CONF_PERIOD_SECS, default=DEFAULT_PERIOD): cv.positive_int,
        vol.Optional(CONF_LOG_SPIKES, default=DEFAULT_LOG_SPIKES): cv.boolean,
        vol.Optional(
            CONF_UPDATE_WHEN_UNAVAILABLE, default=DEFAULT_UPDATE_WHEN_UNAVAILABLE
        ): cv.boolean,
        vol.Optional(CONF_USE_MEDIAN, default=DEFAULT_USE_MEDIAN): cv.boolean,
        vol.Optional(
            CONF_TEMP_RANGE_MIN_CELSIUS, default=DEFAULT_TEMP_RANGE_MIN
        ): float,
        vol.Optional(
            CONF_TEMP_RANGE_MAX_CELSIUS, default=DEFAULT_TEMP_RANGE_MAX
        ): float,
        vol.Optional(
            CONF_TEMPERATURE_ENTITIES, default=DEFAULT_TEMPERATURE_ENTITIES
        ): cv.boolean,
        vol.Optional(
            CONF_HUMIDITY_ENTITIES, default=DEFAULT_HUMIDITY_ENTITIES
        ): cv.boolean,
        vol.Optional(
            CONF_BATTERY_ENTITIES, default=DEFAULT_BATTERY_ENTITIES
        ): cv.boolean,
        vol.Optional(CONF_RSSI_ENTITIES, default=DEFAULT_RSSI_ENTITIES): cv.boolean,
        vol.Optional(
            CONF_NUM_SAMPLES_ENTITIES, default=DEFAULT_NUM_SAMPLES_ENTITIES
        ): cv.boolean,
        vol.Optional(CONF_HCI_DEVICE, default=DEFAULT_HCI_DEVICE): cv.string,
        vol.Optional(CONF_MOAT_DEVICES): vol.All([DEVICES_SCHEMA]),
        vol.Optional(CONF_GOVEE_DEVICES): vol.All([DEVICES_SCHEMA]),
    }
)


###############################################################################


# Entry point!
def setup_platform(
    hass: HomeAssistant, config, add_entities, discovery_info=None
) -> None:
    """Entry point to our integration, called by Home Assistant."""

    _ = discovery_info  # unused, but required by interface

    _LOGGER.info("Starting Bluetooth LE Humidity/Temperature Sensor platform")

    device_wrappers: List[
        SensorDeviceWrapper
    ] = []  # Data objects of configured devices

    from bleson.core.hci import HCIPacket

    def handle_meta_event(hci_packet: HCIPacket) -> None:
        """Handle received BLE data.

        This callback will be called by the HCISocketPoller thread started by the bleson
        library.
        """
        # We only care about BLE packets of type ADVERTISING_REPORT
        if hci_packet.subevent_code == EVT_LE_ADVERTISING_REPORT:
            packet_bd_address = BDAddress(hci_packet.data[3:9])

            for curr_wrapper in device_wrappers:
                # Only process packets from devices that were explicitly added to the HA
                # configuration.
                if BDAddress(curr_wrapper.sensorDevice.mac) == packet_bd_address:
                    _LOGGER.debug(
                        "Received raw_data for %s, len %d (data %d): %s",
                        packet_bd_address,
                        hci_packet.length,
                        hci_packet.length - 11,
                        hex_string(hci_packet.data),
                    )
                    # parse raw_data data
                    parsed_advertisement = parse_ble_advertisement(
                        hci_packet.data, curr_wrapper.brand
                    )

                    # We may parse a model with or without measurements.
                    if parsed_advertisement.model is not None:
                        curr_wrapper.sensorDevice.set_model(parsed_advertisement.model)

                    # If measurements were parsed ("packet" was populated), update
                    # our values.
                    if parsed_advertisement.packet is not None:
                        curr_wrapper.sensorDevice.update(
                            parsed_advertisement.temperature,
                            parsed_advertisement.humidity,
                            parsed_advertisement.battery,
                            parsed_advertisement.battery_millivolts,
                            parsed_advertisement.packet,
                        )

                    # Update RSSI (even if the advertisement didn't have measurement
                    # data)
                    curr_wrapper.sensorDevice.append_rssi(parsed_advertisement.rssi)

    def init_configured_devices() -> None:
        """
        Initialize tracking for all devices listed in our configuration and register
        their Entities to Home Assistant.
        """

        shared_dev_params = CreateDeviceParams(
            report_fahrenheit=config[CONF_REPORT_FAHRENHEIT],
            decimal_places=config[CONF_DECIMALS] if config[CONF_ROUNDING] else None,
            log_spikes=config[CONF_LOG_SPIKES],
            temp_range_min=config[CONF_TEMP_RANGE_MIN_CELSIUS],
            temp_range_max=config[CONF_TEMP_RANGE_MAX_CELSIUS],
        )

        # Initialize a wrapper for each configured device.
        for device_config in config[CONF_MOAT_DEVICES]:
            init_wrapper(device_config, shared_dev_params, DeviceBrand.MOAT)
        for device_config in config[CONF_GOVEE_DEVICES]:
            init_wrapper(device_config, shared_dev_params, DeviceBrand.GOVEE)

    def init_wrapper(
        device_config, device_params: CreateDeviceParams, brand: DeviceBrand
    ):
        # Override the shared configuration with the specific configuration for this
        # device.
        mac = device_config[CONF_DEVICE_MAC]
        device_params.description = device_config.get(CONF_DEVICE_NAME, None)
        device_params.calibrate_temp = device_config.get(
            CONF_DEVICE_CALIBRATE_TEMP, DEFAULT_DEVICE_CALIBRATE_TEMP
        )
        device_params.calibrate_humidity = device_config.get(
            CONF_DEVICE_CALIBRATE_HUMIDITY, DEFAULT_DEVICE_CALIBRATE_HUMIDITY
        )
        new_device = SensorDevice(mac, device_params)
        # Initialize Home Assistant Entities
        name = device_config.get(
            CONF_DEVICE_NAME, mac
        )  # Use provided name, or fall-back to using the MAC.
        ha_entities: List[TempHumSensorEntity] = []
        if config[CONF_TEMPERATURE_ENTITIES]:
            ha_entities.append(
                TemperatureEntity(mac, name, device_params.report_fahrenheit)
            )
        if config[CONF_HUMIDITY_ENTITIES]:
            ha_entities.append(HumidityEntity(mac, name))
        if config[CONF_RSSI_ENTITIES]:
            ha_entities.append(RssiEntity(mac, name))
        if config[CONF_BATTERY_ENTITIES]:
            ha_entities.append(BatteryLevelEntity(mac, name))
        if config[CONF_NUM_SAMPLES_ENTITIES]:
            ha_entities.append(NumSamplesPerPeriodEntity(mac, name))
        new_wrapper = SensorDeviceWrapper(new_device, ha_entities, brand)
        device_wrappers.append(new_wrapper)

        add_entities(ha_entities)

    @callback
    def async_report_device_data() -> None:
        """
        Move the collected data from each SensorDevice to the Home Assistant entities
        and reset the collected data.

        Safe to call within the event loop.
        """
        use_median = config[CONF_USE_MEDIAN]

        num_measurements_attr = "last median of" if use_median else "last mean of"

        for curr_wrapper in device_wrappers:
            entities = curr_wrapper.hassEntities
            device = curr_wrapper.sensorDevice

            temp_num = device.num_measurements
            _LOGGER.debug(
                "Last mfg data for %r: %r, count=%d",
                device.mac,
                device.last_raw_data,
                temp_num,
            )

            curr_entity_index = 0

            # Temperature (optional, based on config)
            if config[CONF_TEMPERATURE_ENTITIES]:
                temperature_mean = device.mean_temperature
                temperature_median = device.median_temperature
                entities[curr_entity_index].device_state_attributes[
                    "mean"
                ] = temperature_mean
                entities[curr_entity_index].device_state_attributes[
                    "median"
                ] = temperature_median
                entities[curr_entity_index].state = (
                    temperature_median if use_median else temperature_mean
                )
                curr_entity_index += 1

            # Humidity (optional, based on config)
            if config[CONF_HUMIDITY_ENTITIES]:
                humidity_mean = device.mean_humidity
                humidity_median = device.median_humidity
                entities[curr_entity_index].device_state_attributes[
                    "mean"
                ] = humidity_mean
                entities[curr_entity_index].device_state_attributes[
                    "median"
                ] = humidity_median
                entities[curr_entity_index].state = (
                    humidity_median if use_median else humidity_mean
                )
                curr_entity_index += 1

            # RSSI (optional, based on config)
            if config[CONF_RSSI_ENTITIES]:
                entities[curr_entity_index].state = device.rssi
                curr_entity_index += 1

            # Battery level (optional, based on config)
            if config[CONF_BATTERY_ENTITIES]:
                entities[curr_entity_index].state = device.battery_percentage
                curr_entity_index += 1

            # Number of samples per period (optional, based on config)
            if config[CONF_NUM_SAMPLES_ENTITIES]:
                entities[curr_entity_index].state = device.num_measurements
                curr_entity_index += 1

            # All entities get the basic [untracked] attributes
            for entity in entities:
                # Skip reporting 'None' if CONF_UPDATE_WHEN_UNAVAILABLE is false.
                # Note that NumSamplesPerPeriodEntity will still be updated,
                # since that is always valid.
                if config[CONF_UPDATE_WHEN_UNAVAILABLE] or entity.state is not None:
                    dev_state_attrs = entity.device_state_attributes
                    dev_state_attrs["last raw data"] = device.last_raw_data
                    dev_state_attrs["rssi"] = device.rssi
                    dev_state_attrs[ATTR_BATTERY_LEVEL] = device.battery_percentage
                    dev_state_attrs["model"] = device.model
                    # Only fill in "battery mV" if the device supports it.
                    # The hasattr case is to make sure we set it to None if it was
                    # previously  filled in, but device.battery_millivolts is None
                    # due to no samples received this period.
                    if (device.battery_millivolts is not None) or hasattr(
                        dev_state_attrs, "battery mV"
                    ):
                        dev_state_attrs["battery mV"] = device.battery_millivolts
                    dev_state_attrs[num_measurements_attr] = device.num_measurements

                    # TODO: even better, don't collect the data if it's disabled.
                    # Check if it's enabled to avoid:
                    #  WARNING (SyncWorker_9) [homeassistant.helpers.entity] Entity
                    #  sensor.deep_freezer_temp is incorrectly being triggered for
                    #  updates while it is disabled. This is a bug in the
                    #  moat_temp_hum_ble integration
                    if entity.enabled:
                        entity.async_schedule_update_ha_state()
                    else:
                        _LOGGER.debug(
                            "Entity %s is disabled; skip update.", entity.entity_id
                        )

            # TODO: There is a race condition.
            # We are in a SyncWorker thread here, reading from "device", but the
            # HCISocketPoller thread can concurrently update "device" (from within
            # handle_meta_event).
            # The impact is that we will occasionally not include a valid measurement
            # when computing the statistics.
            # As long as we're dealing with devices that are sending mostly redundant
            # updates roughly every 2 seconds, I'm not so sure the cost of fixing it
            # (locking or keeping a 2nd copy of the raw measurements data structures)
            # is even worth it.
            if temp_num != device.num_measurements:
                _LOGGER.debug(
                    "Measurements changed from %s to %s!",
                    temp_num,
                    device.num_measurements,
                )

            device.reset()

    def schedule_update_ble_loop() -> None:
        # update_ble_loop() will be called after CONF_PERIOD_SECS elapses.
        # (And in the meantime, the HCISocketPoller thread will make calls to our
        # handle_meta_event callback.)
        point_in_time = dt_util.utcnow() + timedelta(seconds=config[CONF_PERIOD_SECS])
        track_point_in_utc_time(hass, update_ble_loop, point_in_time)  # type: ignore

    def update_ble_loop(now) -> None:
        """
        Update the Home Assistant Entities from our collected data and refresh the
        BLE scanning.
        This probably shouldn't run in the event loop due to start_scanning potentially
        blocking.
        """

        _ = now  # unused, but required by interface

        _LOGGER.debug("update_ble_loop called")

        try:
            # This seems to revive the scanning if someone does:
            # hciconfig hci0 down
            # hciconfig hci0 up
            adapter.start_scanning()

            hass.add_job(async_report_device_data)
        except RuntimeError as error:
            _LOGGER.error("Error during update_ble_loop: %s", error)

        # update_ble_loop() will be called again after CONF_PERIOD_SECS elapses.
        schedule_update_ble_loop()

    ###########################################################################

    # Initialize bluetooth adapter and begin scanning.
    # XXX: only supports LinuxProvider implementation (other platforms don't have
    #   _handle_meta_event to monkey-patch).
    # XXX: only supports hci0 - hci9
    adapter: Adapter = get_provider().get_adapter(int(config[CONF_HCI_DEVICE][-1]))
    # Monkey-patch the BluetoothHCIAdapter.  Would be better if the Adapter interface
    # exposed this functionality.
    setattr(adapter, "_handle_meta_event", handle_meta_event)
    hass.bus.listen("homeassistant_stop", adapter.stop_scanning)
    adapter.start_scanning()

    # Initialize configured devices
    init_configured_devices()

    # update_ble_loop() will first be called after CONF_PERIOD_SECS elapses.
    schedule_update_ble_loop()

    _LOGGER.info("Done setting up Bluetooth LE Humidity/Temperature Sensor platform")


class TempHumSensorEntity(Entity):
    """Base class for our HomeAssistant Entity classes."""

    def __init__(
        self, mac: str, entity_name: str, device_name: str, unique_id_prefix: str
    ):
        """Initialize the Entity."""
        self._state: Union[None, str, int, float] = None
        self._battery = None
        self._mac = mac
        self._unique_id = f'{unique_id_prefix}_{mac.replace(":", "")}'
        self._entity_name = entity_name
        self._device_name = device_name
        self._device_state_attributes: Dict[str, Any] = {}

    @property
    def name(self) -> str:
        """Return the UI-friendly name of the Entity."""
        return self._entity_name

    @property
    def state(self) -> Union[None, str, int, float]:
        """Return the state of the Entity."""
        return self._state

    @state.setter
    def state(self, state: Union[None, str, int, float]) -> None:
        """Set the state of the Entity."""
        self._state = state

    @property
    def should_poll(self) -> bool:
        """
        Don't poll; we will call schedule_update_ha_state from within our periodic
        update_ble_loop.
        """
        return False

    @property
    def device_state_attributes(self) -> Dict[str, Any]:
        """Return the state attributes."""
        return self._device_state_attributes

    @property
    def unique_id(self) -> str:
        """Return the persistent unique ID."""
        return self._unique_id

    @property
    def force_update(self) -> bool:
        """Force update."""
        return True

    @property
    def device_info(self) -> Optional[Dict[str, Any]]:
        return {
            "identifiers": {(sensor.DOMAIN, self._mac)},
            "name": self._device_name,
        }

    @property
    def available(self) -> bool:
        """Return whether the device is available or not."""
        return self.state is not None

    async def async_added_to_hass(self) -> None:
        """Run when entity about to be added to hass."""
        _LOGGER.info(
            "Adding entity %s (%s): %s", self.entity_id, self.unique_id, self.name
        )

    async def async_will_remove_from_hass(self) -> None:
        """Run when entity will be removed from hass."""
        _LOGGER.info(
            "Removing entity %s (%s): %s", self.entity_id, self.unique_id, self.name
        )


class TemperatureEntity(TempHumSensorEntity):
    """HomeAssistant Entity for tracking temperature."""

    def __init__(self, mac: str, device_name: str, report_fahrenheit: bool):
        TempHumSensorEntity.__init__(
            self,
            mac=mac,
            device_name=device_name,
            entity_name=f"{device_name} temp",
            unique_id_prefix="t",
        )
        self._report_fahrenheit = report_fahrenheit

    @property
    def unit_of_measurement(self) -> str:
        """Return the unit of measurement."""
        return TEMP_FAHRENHEIT if self._report_fahrenheit else TEMP_CELSIUS

    @property
    def device_class(self):
        """Return the Home Assistant device class."""
        return DEVICE_CLASS_TEMPERATURE


class HumidityEntity(TempHumSensorEntity):
    """HomeAssistant Entity for tracking humidity."""

    def __init__(self, mac: str, device_name: str):
        TempHumSensorEntity.__init__(
            self,
            mac=mac,
            device_name=device_name,
            entity_name=f"{device_name} humidity",
            unique_id_prefix="h",
        )

    @property
    def unit_of_measurement(self) -> str:
        """Return the unit of measurement."""
        return "%"

    @property
    def device_class(self):
        """Return the Home Assistant device class."""
        return DEVICE_CLASS_HUMIDITY


class RssiEntity(TempHumSensorEntity):
    """HomeAssistant Entity for tracking RSSI."""

    def __init__(self, mac: str, device_name: str):
        TempHumSensorEntity.__init__(
            self,
            mac=mac,
            device_name=device_name,
            entity_name=f"{device_name} RSSI",
            unique_id_prefix="r",
        )

    @property
    def unit_of_measurement(self) -> str:
        """Return the unit of measurement."""
        return "dBm"

    @property
    def device_class(self):
        """Return the Home Assistant device class."""
        return DEVICE_CLASS_SIGNAL_STRENGTH


class BatteryLevelEntity(TempHumSensorEntity):
    """HomeAssistant Entity for tracking battery level."""

    def __init__(self, mac: str, device_name: str):
        TempHumSensorEntity.__init__(
            self,
            mac=mac,
            device_name=device_name,
            entity_name=f"{device_name} battery level",
            unique_id_prefix="b",
        )

    @property
    def unit_of_measurement(self) -> str:
        """Return the unit of measurement."""
        return "%"

    @property
    def device_class(self):
        """Return the Home Assistant device class."""
        return DEVICE_CLASS_BATTERY


class NumSamplesPerPeriodEntity(TempHumSensorEntity):
    """
    HomeAssistant Entity for tracking number of the number measurements made during a
    period of time.
    """

    def __init__(self, mac: str, device_name: str):
        TempHumSensorEntity.__init__(
            self,
            mac=mac,
            device_name=device_name,
            entity_name=f"{device_name} samples",
            unique_id_prefix="s",
        )

    @property
    def unit_of_measurement(self) -> Optional[str]:
        """Return the unit of measurement."""
        return "samples/period"

    @property
    def device_class(self):
        """Return the Home Assistant device class."""
        return None


@dataclass
class SensorDeviceWrapper:
    sensorDevice: SensorDevice
    hassEntities: Sequence[TempHumSensorEntity]
    brand: DeviceBrand


custom_components/moat_temp_hum_ble/const.py
"""Constants for the BLE HCI monitor sensor integration."""

from enum import Enum

# Configuration options
CONF_REPORT_FAHRENHEIT = "report_fahrenheit"
CONF_ROUNDING = "rounding"
CONF_DECIMALS = "decimals"
CONF_PERIOD_SECS = "period_secs"
CONF_LOG_SPIKES = "log_spikes"
CONF_UPDATE_WHEN_UNAVAILABLE = "update_when_unavailable"
CONF_USE_MEDIAN = "use_median"
CONF_TEMP_RANGE_MIN_CELSIUS = "temp_range_min_celsius"
CONF_TEMP_RANGE_MAX_CELSIUS = "temp_range_max_celsius"
CONF_TEMPERATURE_ENTITIES = "temperature_entities"
CONF_HUMIDITY_ENTITIES = "humidity_entities"
CONF_BATTERY_ENTITIES = "battery_entities"
CONF_NUM_SAMPLES_ENTITIES = "num_samples_entities"
CONF_RSSI_ENTITIES = "rssi_entities"
CONF_HCI_DEVICE = "hci_device"
CONF_MOAT_DEVICES = "moat_devices"
CONF_GOVEE_DEVICES = "govee_devices"
CONF_DEVICE_MAC = "mac"
CONF_DEVICE_NAME = "name"
CONF_DEVICE_CALIBRATE_TEMP = "calibrate_temp"
CONF_DEVICE_CALIBRATE_HUMIDITY = "calibrate_humidity"

# Default values for configuration options
DEFAULT_REPORT_FAHRENHEIT = False
DEFAULT_ROUNDING = True
DEFAULT_DECIMALS = 2
DEFAULT_PERIOD = 60
DEFAULT_LOG_SPIKES = True
DEFAULT_UPDATE_WHEN_UNAVAILABLE = True
DEFAULT_USE_MEDIAN = False
DEFAULT_TEMP_RANGE_MIN = -45.0
DEFAULT_TEMP_RANGE_MAX = 70.0
DEFAULT_TEMPERATURE_ENTITIES = True
DEFAULT_HUMIDITY_ENTITIES = True
DEFAULT_BATTERY_ENTITIES = False
DEFAULT_RSSI_ENTITIES = False
DEFAULT_NUM_SAMPLES_ENTITIES = False
DEFAULT_HCI_DEVICE = "hci0"
DEFAULT_DEVICE_CALIBRATE_TEMP = 0.0
DEFAULT_DEVICE_CALIBRATE_HUMIDITY = 0.0

# Fixed constants:

# Sensor measurement limits to exclude erroneous spikes from the results
HUMIDITY_MIN = 0.0
HUMIDITY_MAX = 99.9


class DeviceBrand(Enum):
    MOAT = 1
    GOVEE = 2
